pubsub/pumps.go

```diff
@@ -0,0 +1,127 @@
+package pubsub
+
+import (
+	"net"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+// readPump tries to read out data from the connection, sending it down
+// the data channel, until it's closed. It's a conversion from
+// synchronous to channel-based code.
+type readPump struct {
+	cnx redis.Conn
+
+	data   chan interface{}
+	errs   chan error
+	closer chan struct{}
+}
+
+// newReadPump creates a new pump that operates on the single Redis connection.
+func newReadPump(cnx redis.Conn) *readPump {
+	return &readPump{
+		cnx:    cnx,
+		data:   make(chan interface{}),
+		errs:   make(chan error),
+		closer: make(chan struct{}),
+	}
+}
+
+// Work starts reading from the connection and blocks until it is closed.
+func (r *readPump) Work() {
+	cnx := redis.PubSubConn{Conn: r.cnx}
+	defer close(r.closer)
+
+	for {
+		msg := cnx.Receive()
+		err, isErr := msg.(error)
+
+		if !isErr {
+			select {
+			case r.data <- msg:
+			case <-r.closer:
+				return
+			}
+		} else if shouldNotifyUser(err) {
+			select {
+			case r.errs <- err:
+			case <-r.closer:
+				return
+			}
+		}
+	}
+}
+
+// Errs returns a channel of errors from the connection
+func (r *readPump) Errs() <-chan error { return r.errs }
+
+// Data returns a channel of pubsub data from the connection
+func (r *readPump) Data() <-chan interface{} { return r.data }
+
+// Close tears down the connection
+func (r *readPump) Close() { r.closer <- struct{}{} }
+
+type command struct {
+	command string
+	channel string
+	written chan<- struct{}
```

Comment:
I think

```go
<-written
return
```

is preferable to

```go
c.cond.L.Lock()
defer c.cond.L.Unlock()
for {
    if c.written {
        return
    }
    c.cond.L.Unlock()
}
```

If you look at it, many channels (like closer channels) are one-time uses. They're used for signaling, which is what we need here.